packages/compiler/auto.ts

```diff
@@ -0,0 +1,306 @@
+      t.FunctionExpression | t.ArrowFunctionExpression
+    >;
+  } else {
+    componentPath = path as NodePath<t.FunctionDeclaration>;
+  }
+
+  if (!t.isIdentifier(rawComponent.id)) return;
+  if (!isComponent(rawComponent.id.name)) return;
+
+  const globalPath = path.findParent(
+    (path) => path.parentPath?.isProgram() || path.isProgram()
+  )!;
+  const comment =
+    rawComponent.leadingComments?.[0] ||
+    path.parent.leadingComments?.[0] ||
+    globalPath.node.leadingComments?.[0];
+
+  if (comment?.value.includes('million-ignore')) {
+    logIgnore(rawComponent.id.name);
+    return;
+  }
+
+  const blockStatementPath = componentPath.get(
+    'body'
+  ) as NodePath<t.BlockStatement>;
+  const returnStatementPath = blockStatementPath
+    .get('body')
+    .find((path) => path.isReturnStatement());
+
+  if (!returnStatementPath?.has('argument')) return;
+  const argumentPath = returnStatementPath.get('argument') as NodePath<
+    t.JSXElement | t.JSXFragment
+  >;
+
+  if (!argumentPath.isJSXElement() && !argumentPath.isJSXFragment()) {
+    return;
+  }
+
+  const componentInfo = {
+    bailout: false,
+    elements: 0,
+    attributes: 0,
+    components: 0,
+    text: 0,
+    expressions: 0,
+    depth: 0,
+    returns: 0,
+  };
+
+  const skipIds: (string | RegExp)[] = [];
+
+  if (typeof info.options.auto === 'object' && info.options.auto.skip) {
+    skipIds.push(...info.options.auto.skip);
+  }
+
+  blockStatementPath.traverse({
+    Identifier(path) {
+      if (
+        skipIds.some((id) =>
+          typeof id === 'string'
+            ? path.node.name === id
+            : id.test(path.node.name)
+        )
+      ) {
+        componentInfo.bailout = true;
+        path.stop();
+        return;
+      }
+    },
+    JSXElement(path) {
+      const type = path.node.openingElement.name;
+      if (t.isJSXMemberExpression(type)) {
+        const isContext =
+          t.isJSXIdentifier(type.object) &&
+          isComponent(type.property.name) &&
+          type.property.name === 'Provider';
+        const isSpecialElement = t.isJSXIdentifier(type.object);
+
+        if (isContext) {
+          componentInfo.bailout = true;
+          path.stop();
+          return;
+        }
+
+        if (isSpecialElement) {
+          componentInfo.components++;
+          path.skip();
+          return;
+        }
+      }
+      if (t.isJSXIdentifier(type) && isComponent(type.name)) {
+        componentInfo.components++;
+        path.skip();
+        return;
+      }
+
+      componentInfo.elements++;
+    },
+    JSXSpreadChild(path) {
+      componentInfo.components++;
+      path.skip();
+    },
+    JSXExpressionContainer(path) {
+      if (t.isJSXEmptyExpression(path.node.expression)) return;
+
+      if (
+        t.isCallExpression(path.node.expression) &&
+        t.isMemberExpression(path.node.expression.callee) &&
+        t.isIdentifier(path.node.expression.callee.property, { name: 'map' })
```

Comment:
This has holes, but whatever, could be improved. (Same goes to other bruteforce checks)